### PR TITLE
Reduce CLP to the French Five

### PR DESCRIPTION
**Pull request by michielbdejong**

### CODE REVIEW

**Comment by michielbdejong** [comment]
Rebased on current version of #263

File: asn1/CommonLedgerProtocol.asn [hunk 1]
```diff
@@ -21,22 +21,24 @@ SideProtocolData ::= SEQUENCE OF SEQUENCE {
   protocolData OCTET STRING
 }
 
-Ack ::= SEQUENCE {
-  ack NULL,
+Request ::= SEQUENCE {
+  packet CHOICE {
```
Comment (emschwartz) [inline asn1/CommonLedgerProtocol.asn]: As commented in https://github.com/interledger/rfcs/pull/263#issuecomment-320779560, I think it's better to add an additional type than send an extra byte over the wire for every response.

Comment (adrianhopebailie) [inline asn1/CommonLedgerProtocol.asn]: +1 - Having lots of types in the definition is not a bad thing.

**Comment by adrianhopebailie** [comment]
I believe there is a layering issue in the set of generic packets that is making this harder than it should be. If you look at this list, the only packet that is relayed, unchanged, between nodes is the `InterledgerProtocolPayment`.

```
    {1 InterledgerProtocolPayment} |
    {2 QuoteLiquidityRequest} |
    {3 QuoteLiquidityResponse} |
    {4 QuoteBySourceAmountRequest} |
    {5 QuoteBySourceAmountResponse} |
    {6 QuoteByDestinationAmountRequest} |
    {7 QuoteByDestinationAmountResponse} |
    {8 InterledgerProtocolError}
```

This suggests that the `InterledgerProtocolPayment` should exist one layer higher in the stack and be encapsulated in a packet that is passed from node to node like the others.

We should remove `InterledgerProtocolPayment` and add `TransferRequest`, `SuccessfulTransferResponse` and `FailedTransferResponse` defined as follows:

```
TransferRequest ::= SEQUENCE {
  transferId UInt128,
  amount UInt64,
  executionCondition UInt256,
  expiresAt Timestamp,
  packet InterledgerProtocolPayment,
}

SuccessfulTransferResponse ::= SEQUENCE {
  transferId UInt128,
  fulfillment UInt256
}

FailedTransferResponse ::= SEQUENCE {
  transferId UInt128,
  reason InterledgerProtocolError
}
```

It's not clear if the ILQP requests can ever have an ILP Error response so let's assume that is only used in response to a transfer request.

Therefor the new set of Generic Packets are:  
```
    {1 TransferRequest} |
    {2 SuccessfulTransferResponse} |
    {3 FailedTransferResponse} |
    {4 QuoteLiquidityRequest} |
    {5 QuoteLiquidityResponse} |
    {6 QuoteBySourceAmountRequest} |
    {7 QuoteBySourceAmountResponse} |
    {8 QuoteByDestinationAmountRequest} |
    {9 QuoteByDestinationAmountResponse} 
```
This is the protocol layer that exists below ILP, the messages exchanged directly between nodes (connectors, senders and receivers). 

The transfer request encapsulates the `InterledgerProtocolPayment` (often referred to as the ILP Packet) along with the other required data for a transfer (expiry, condition etc).

Note that the `...Request` and `...Response` at this layer imply the logical flow not the message flow (which is captured in the next layer down, the CommonMessageProtocol). For example, a `TransferRequest` may be carried in a `RequestMessage` and the `ResponseMessage` may simply carry an `Ack`. Then the `SuccessfulTransferResponse` is returned, carried in a `RequestMessage` envelope, because it is initiated from the upstream node. 

The CommonMessageProtocol could then be reduced to:

```

Ack ::= SEQUENCE {
  -- The messageId of the request
  correlationId UInt32,
  -- Additional data for protocol extensibility
  sideProtocolData SideProtocolData
}

RequestMessage ::= SEQUENCE {
  packet InterledgerPacket,
  -- Additional data for protocol extensibility
  sideProtocolData SideProtocolData
}

ResponseMessage ::= SEQUENCE {
  -- The messageId of the request
  correlationId UInt32,
  packet InterledgerPacket,
  -- Additional data for protocol extensibility
  sideProtocolData SideProtocolData
}

CustomRequestMessage ::= SEQUENCE {
  sideProtocolData SideProtocolData
}

CustomResponseMessage ::= SEQUENCE {
  correlationId UInt32,
  sideProtocolData SideProtocolData
}

CALL ::= CLASS {
    &typeId UInt8 UNIQUE,
    &Type
} WITH SYNTAX {&typeId &Type}

CallSet CALL ::= {
    {1 Ack} |
    {2 RequestMessage} |
    {3 ResponseMessage} |
    {4 CustomRequestMessage} |
    {5 CustomResponseMessage} 
}

CommonMessageProtocolPacket ::= SEQUENCE {
    -- One byte type ID
    type CALL.&typeId ({CallSet}),
    -- Used to associate requests and corresponding responses
    -- If messageId = 0, the server MUST not send a response
    messageId UInt32,
    -- Length-prefixed main data
    data CALL.&Type ({CallSet}{@type})
}
```

**Comment by michielbdejong** [comment]
> I believe there is a layering issue in the set of generic packets that is making this harder than it should be.

I had exactly the same reaction on Friday. Evan started explaining why he wants to have it layered with separate "ILP-packet" layer and "ledger layer".

I had forgotten about this point, but now that Adrian brought it up independently, I agree that it would be simpler make the CLP layer simpler if it just deals with request-response and side-data. 

@emschwartz can you explain again why you don't want to put the prepare/fulfill/reject types next the quote/response/ILP-error types, in the GenericPacket registry?

**Comment by michielbdejong** [comment]
I added back the TransferProtocol at the GenericPacket layer and now the CLP is really, really simple, just deals with linking responses to requests, and with request/response types that just predict whether it has a packet and/or SideData.

**Comment by emschwartz** [comment]
> I believe there is a layering issue in the set of generic packets that is making this harder than it should be. If you look at this list, the only packet that is relayed, unchanged, between nodes is the InterledgerProtocolPayment.

That is not true. Quotes can be relayed as well (the amounts may or may not be changed but the message format stays the same) and errors definitely are.

> It's not clear if the ILQP requests can ever have an ILP Error response

The definitely can! Why wouldn't they? If you can't get a response, why not respond with an `F02: Unreachable` error?

This is turning into a discussion of how we think about the protocol layers so I'm going to respond to the general question in https://github.com/interledger/rfcs/issues/253.

**Comment by adrianhopebailie** [comment]
> the amounts may or may not be changed but the message format stays the same

If anything changes it's not an end-to-end message like the InterledgerProtocolPayment. They shouldn't be in the same layer.

> If you can't get a response, why not respond with an F02: Unreachable error?

Encapsulated how? How does the sender know that the error corresponds to the request?

I have a counter-proposal to these designs which I'll put in a PR

**Comment by emschwartz** [comment]
The Interledger layer is not meant to be end-to-end, it's the part that's understood and implemented by _every_ party in the path. Even in the case of the InterledgerProtocolPayment the address and amount are used by the connectors throughout the path. End-to-end means that the data is irrelevant and should be treated as completely opaque by everyone other than the sender and receiver. The Application and Transport layers, as we have currently defined them, are end-to-end.

In whatever Ledger layer protocol you use to talk to a connector we assume that there is a request/response flow (currently the quote requests and responses have IDs in them for correlation but we are leaning more towards deprecating those fields now). That's how you get the error back.

**Comment by michielbdejong** [comment]
I still feel this way of organizing "CLP does the request/response, GenericPacket does the quoting/transfers" would have been more logical, and I think it would have been easier to understand for newcomers, but if you define CLP as a virtual ledger that supports Interledger, then Evan's latest proposal, with which Ben already agrees, is also defendable.

In the end, my main interest in this is to hope this is the last one in a series of rapid changes, and once we agree on CLP, we can enter relatively stable waters. I would still like to propose a freeze once this is merged, of several months where we don't change any of protocols.

This PR is now superseded by #267, which contains just a few minor edits to Evan's proposal.

**Pull request closed without merging**